examples/water_sampling_common.py

```diff
@@ -0,0 +1,166 @@
+    dijs = np.linalg.norm(diffs, axis=-1)
+    count = np.sum(dijs < threshold)
+    return count
+
+
+def get_initial_state(water_pdb, mol, ff, seed, nb_cutoff, lamb):
+    assert nb_cutoff == 1.2  # hardcoded in prepare_host_edge
+
+    # read water system
+    solvent_sys, solvent_conf, solvent_box, solvent_topology, num_water_atoms = build_system(
+        water_pdb, ff.water_ff, padding=0.1
+    )
+    solvent_box += np.diag([0.1, 0.1, 0.1])  # remove any possible clashes
+
+    assert num_water_atoms == len(solvent_conf)
+
+    bt = BaseTopology(mol, ff)
+    afe = AbsoluteFreeEnergy(mol, bt)
+    host_config = HostConfig(solvent_sys, solvent_conf, solvent_box, num_water_atoms)
+    potentials, params, combined_masses = afe.prepare_host_edge(ff.get_params(), host_config, lamb)
+
+    ligand_idxs = np.arange(num_water_atoms, num_water_atoms + mol.GetNumAtoms())
+
+    # modify state with all ligand atoms decoupled (returned by prepare_host_edge) to
+    # state with only half of the ligand atoms decoupled. This roughly decouples a contingous
+    # semi-sphere along the "wide" axis as opposed to form random small pores.
+    nb_params = np.array(params[-1])
+
+    final_masses = combined_masses
+    final_conf = np.concatenate([solvent_conf, get_romol_conf(mol)], axis=0)
+
+    component_dim = 4  # q,s,e,w
+    num_atoms = len(final_conf)
+
+    ligand_water_flat_idxs = np.s_[component_dim * num_atoms : 2 * component_dim * num_atoms]  # water
+    ligand_water_params = nb_params[ligand_water_flat_idxs].reshape(-1, 4)
+    # forms a hole on the "flat" side of the buckyball, 1-indexed
+    # previously failed attempts:
+    # 1) uniformly decoupling half of the buckyball
+    # 2) uniformly decoupling entire buckyball
+    # 3) uniformly shrinking vdw radii of every buckyball atom
+    # 4) uniformly shrinking vdw radii of half of the buckyball atoms
+    lambda_coupled_ligand_atoms = num_water_atoms + np.array([10, 11, 16, 17, 21, 22, 12, 14, 13, 25, 24, 23, 20]) - 1
+    fully_coupled_ligand_atoms = np.array(
+        list(set(ligand_idxs.tolist()).difference(lambda_coupled_ligand_atoms.tolist()))
+    )
+
+    ligand_water_params[fully_coupled_ligand_atoms, -1] = 0
+    nb_params[ligand_water_flat_idxs] = ligand_water_params.reshape(-1)
+
+    # override last potential with the updated nb_params
+    host_bps = []
+    for p, bp in zip(params[:-1], potentials[:-1]):
+        host_bps.append(bp.bind(p))
+
+    host_bps.append(potentials[-1].bind(nb_params))
+
+    # (YTZ): This is disabled because the initial ligand and starting waters are pre-minimized
+    # print("Minimizing the host system...", end="", flush=True)
+    # host_conf = minimize_host_4d([mol], host_config, ff)
+    # print("Done", flush=True)
+
+    temperature = DEFAULT_TEMP
+    dt = 1.0e-3
+    barostat_interval = 25
+    integrator = LangevinIntegrator(temperature, dt, 1.0, final_masses, seed)
+    bond_list = get_bond_list(host_bps[0].potential)
```

Comment:
nit: do we want to move away from assuming an ordering of the potentials?
```suggestion
    bond_list = get_bond_list(next(bp for bp in host_bps if isinstance(bp, HarmonicBond)).potential)
```

Reply:
Sorry, just noticed that my suggested edit was wrong. I think it should be

```python
bond_list = get_bond_list(next(bp for bp in host_bps if isinstance(bp.potential, HarmonicBond)).potential)
```